### ISSUE TITLE
is it possible to merge text operations?

### ISSUE THREAD

**Issue by calibr** [y-js/yjs#136]
Now in YText when I do this:

```
text.applyDelta([{ insert: 'hello' }])
```

YJS creates one operation with content 'hello'

But in a real life text editor this will look like a sequence of one character insertions when a user types something:

```
text.applyDelta([{ insert: 'h' }])
```

```
text.applyDelta([{ retain: 1 }, { insert: 'e' }])
```

```
text.applyDelta([{ retain: 2 }, { insert: 'l' }])
```

etc...

and this creates 5 operations and takes more memory because of the operations meta data.

I made a test which makes many of one character insertions and here is the memory rate graph:

<img width="917" alt="typing_graph_no_titles" src="https://user-images.githubusercontent.com/4965093/50426218-f7fb3f00-0898-11e9-9043-f5c9a787d242.png">

(x axis - number of symbols inserted, y - consumed memory)

So for 1M symbols a user inserted YJS will probably take something like 200M of RAM.

Here I made `applyDelta` calls each with a 10kb string, and memory consumptions looks much better:

<img width="908" alt="graph_no_titles_2kb_op_size" src="https://user-images.githubusercontent.com/4965093/50426221-2416c000-0899-11e9-9770-4b68c68a4ae5.png">

Is there is a plan to compact sequential text inserts somehow, or maybe it is already possible to do that?

**Comment by dmonad** [comment]
When text is written sequentially (v12 - the version you are using) should actually merge operations.

So when you type the text "hello" the operations are merged into one operation.

This does not work when the text is written in non-sequential order (e.g. right to left).

I disabled that feature partially in the current Yjs version (v13 - beta), because it conflicts with how the undo manager works. But I think we should enable it again when no undo manager is used.

-- 

I did a complete rewrite of Yjs in v13 and it uses memory much more efficiently. The above issue won't persist in the current version.

I'm sorry to say that I'm lacking the time to provide support for v12 and fix this issue..

**Comment by calibr** [comment]
Hello  @dmonad  thank you for the response. Actually I use v13, sorry I should have mention that because it is not obvious from the code I sent.

We use UndoManager in our text editor and it works perfectly fine so far(I a bit extended it to save  arbitrary data like caret position etc. with undo records). So probably we need to find another way to reduce the memory usage.

In our app we use client/server architecture with the WebSocket provider now and the memory growing can affect our servers, rather than clients. So I want to try using Ydb, it looks very promising, but as I understand the Ydb provider is not completely ready to use - it uses wrong paths for the deps(like `../MessageHandler/integrateRemoteStructs.js`). I hope this is the only issue with it and I will be able to test it with a little fixing. Cheers!

**Comment by dmonad** [comment]
The Ydb project is idle at the moment. I'm currently looking for a sponsor, because the project is pretty ambitious. 

Can you try calling `defragmentItemContent(y)`? - This will merge all operations, if possible. You find it in `./utils/defragmentItemContent.js`

If this fixes the problem, I will provide an option to enable merging on the server.

**Comment by calibr** [comment]
@dmonad yes it helped. During the tests defragmentItemContent is being called every second and at the end a doc with 1M symbols takes ~50Mb(x4 better than before). When I reload this doc from levelDB it takes just a few megabytes of memory, so probably memory leeks somewhere even after `defragmentItemContent`(or maybe it is some additional modules like UndoManager take this memory?). I will look into this more thoroughly to understand what happens with the memory(maybe problem somewhere in my tests) and will update you as soon as I have more info.

**Comment by calibr** [comment]
The problem was with the string concatenation in defragmentItemContent, I made a quick and dirty modification to use a temp buffer https://github.com/calibr/yjs/blob/master/utils/defragmentItemContent.js#L33 and it consumes roughly the same amount of memory as here https://user-images.githubusercontent.com/4965093/50426221-2416c000-0899-11e9-9770-4b68c68a4ae5.png, that is very good. If you are interested you can find more info about the V8 string concatenation here https://twitter.com/mourner/status/973664460291411969?lang=en, actually it is about speed, but it turns out memory usage is also affected if strings are concatenated symbol by symbol.

About the option that would enable merging on the server, how would it work? It looks like merging ops is quite expensive because it traverses all the nodes in a  document, so I am wondering how often it should be called..

**Comment by dmonad** [comment]
Thanks for digging into this. Before I wrote the code, I made some research on the fasted method to concatenate strings (e.g. https://stackoverflow.com/questions/16696632/most-efficient-way-to-concatenate-strings-in-javascript). 

Also tests show to verify that += is faster. But when you only append a character at a time, the join method seems to be the fastest.

When you inspect the proposed methods in the chromium performance panel, you can see that the garbage collector turns on several thousand times in the buffer and += methods. Meanwhile the join method invokes the garbage collector only once. 

I will switch to the join method. Conceptually this makes the most sense to me.

Regarding the merge method: This should only be called on Items that actually were modified. You can observe the Items/Types that changed after each transaction, and try to merge each item individually.

```
y.on('afterTrasaction', transaction => {
  transaction.changedTypes
  transaction.removedStructs
...
})
```

Again, I don't want to add this feature now, because I'm working on Undo/Redo and Snapshots - and this conflicts with my work. But I will eventually re-introduce automatic merging.

**Comment by dmonad** [comment]
Yjs just got a major refactor. v13.0.0-80 now efficiently merges automatically after each transaction.
